### ISSUE TITLE
Duplicate JSON files provided to OPA could be dealt with more gracefully

### ISSUE THREAD

**Issue by anderseknert** [open-policy-agent/opa#2983]
If a JSON file is provided more than once when starting OPA, a merge error is thrown. This came up on Slack when a user mistakenly used something like `opa test . input.json` and got a cryptic merge error as a result.

## Expected Behavior

OPA should be able to either 1) filter out duplicates from the list of provided files or 2) give a more specific error message if duplicates are found.

## Actual Behavior

`1 error occurred during loading: data.json: merge error`

## Steps to Reproduce the Problem
```
$ echo '{"a":"b"}' > data.json
$ opa test . data.json
```

**Comment by faraway** [comment]
@anderseknert just saw you added the "good first issue" label. Do you think this is something I can take can fix?

**Comment by anderseknert** [comment]
@faraway I think @olamiko might be working on this - seems like I can't assign issues unless that person has interacted with an issue first, but I know I tried. Are you on the OPA Slack? We could sync there if you're looking for a good first issue to tackle.

**Comment by olamiko** [comment]
Hi @faraway, I am currently working on this

**Comment by anderseknert** [comment]
This came up again today, when someone ran something like `opa check src/ test/ .` and saw merge errors.

**Comment by renatosc** [comment]
I am seeing this error when trying to do `opa test .` (from my `src' folder) which is giving me

`1 error occurred during loading: src\schemas\check1_schema.json`

I do not have duplicated files inside the src folder and I started getting this error when I added a 2nd schema json file to my schema folder.

**Comment by anderseknert** [comment]
OPA merges all data it encounters according to its [document model](https://www.openpolicyagent.org/docs/latest/philosophy/#the-opa-document-model). You don't need to have duplicate **files** for this error to be displayed — just data that exists at the same path in two different JSON documents. I.e.

**foo.json**
```json
{"foo": 1}
```

**bar.json**
```json
{"foo": 2}
```
Would be a conflict, as foo will be merged into the same document from two different sources.

**Comment by anderseknert** [comment]
I believe some attribute is merged with a conflict :) If you can create a minimal repo to reproduce I'd be happy to take a look.

**Comment by renatosc** [comment]
There you go: https://github.com/renatosc/opa-merge-error

**Comment by anderseknert** [comment]
Thanks Renato! That **is** a conflict, as both of your schemas contain the same keys at the same level of the document. You'd either need to put each under it's own top-level key, like:

```json
{
  "schema1": { ... }
}
```
Or use a directory structure where each file is in its own directory, and treat the topmost directory as a bundle (i.e. `--bundle`) in which case the directory structure will be mirrored in the document.

**Comment by renatosc** [comment]
Hi @anderseknert, thanks for checking it out. 

I tried your suggestion of adding a key (like as "**schema1**") before the current schema json object and although it fixes the merge error it appears to make the schema validation to not work properly. 

I updated the repo by updating the policy1_schema.json to have a "**schema1**" key as you suggested. I also changed the schema to expect a `role2` property which currently does not exist in policy1.rego (so we can see the schema validation working).

So, by running

`../opa eval data.policy1.allow --data regos/policy1.rego --schema schemas/policy1_schema.json`

I am now getting no error back when I should in fact be receiving an error (since schema is expecting `role2` and rego is using  `rego`.)

If I remove that "schema1" key added the schema file and run the command above, I get the schema validation error correctly.

**Comment by anderseknert** [comment]
Ah okay, I think I see what's going on here. OPA merges all documents that are included for *policy evaluation* — i.e. all documents that contribute to the **data** document. The documents loaded using the `--schema` flag do not contribute directly to the data document, but are only used as schemas, while running e.g. `opa test .` naturally will include **all** documents in the directory, includsing the `schemas` subdirectory but treat them as **data**, not as schemas. Does that make any sense?

**Comment by renatosc** [comment]
Yes, which goes to my point that maybe we could have a bug on `opa test`. Want me to take a look on that or do you think it may be too complex for opa first-time contributor to take care of?

**Comment by anderseknert** [comment]
Not sure what the bug would be though. If you run `opa test .` you're explicitly asking it to use **all** files in the directory, including in any subdirectories. In that context a JSON schema is just any other JSON file.

Rather than running `opa test .`, I'd suggest keeping your policy in a separate directory, e.g.

```
.
├── data
│   └── data1.json
├── policies
│   ├── policy1.rego
│   └── policy2.rego
└── schemas
    ├── schema1.json
    └── schema2.json
```

And then you'll be able to run `opa test policy/` to run your tests, and `opa eval -d data/ -d policies/ --schema schemas` to run evaluation. 

If you'd like to contribute, perhaps adding the `--schema` flag to `opa test` would be a good candidate? I can't find an issue for it, but I know it's been requested in the past.

**Comment by renatosc** [comment]
I think adding a `--schema` flag to `opa test` makes sense and keeps it consistent with `opa eval`.

Right now I was able to make it work by calling `opa test --ignore schemas`, but having the `--schema` param and making `opa test` also validate the schema I think would be a nice addition (for instance, right now I call `opa test` for unit test and then `opa eval` for schema validation. If `opa test` did both I do not need to call `opa eval` anymore)

**Comment by anderseknert** [comment]
Agreed, that'd be a nice addition. @johanfylling do you remember where this was discussed in the past?
I doubt that anyone objects to this, but if you do, now would be a good time to say so :)

**Comment by johanfylling** [comment]
@anderseknert , this is probably the issue you're thinking of: https://github.com/open-policy-agent/opa/issues/5430

**Comment by anderseknert** [comment]
Thanks @johanfylling! That was it, yes. @renatosc if you'd like to work on that, I'd be happy to create an issue.
